runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/util/PackageUtil.java

```diff
@@ -217,69 +225,99 @@ private StagingResult tryStagePackageWithRetry(
                       + "of %s. Please verify credentials are valid and that you have "
                       + "write access to %s. Stale credentials can be resolved by executing "
                       + "'gcloud auth application-default login'.",
-                  source, target);
+                  sourceDescription, target);
           LOG.error(errorMessage);
           throw new IOException(errorMessage, ioException);
         }
 
         long sleep = backoff.nextBackOffMillis();
         if (sleep == BackOff.STOP) {
-          LOG.error("Upload failed, will NOT retry staging of package: {}", source, ioException);
+          LOG.error(
+              "Upload failed, will NOT retry staging of package: {}",
+              sourceDescription,
+              ioException);
           throw new RuntimeException("Could not stage %s to %s", ioException);
         } else {
           LOG.warn(
               "Upload attempt failed, sleeping before retrying staging of package: {}",
-              source,
+              sourceDescription,
               ioException);
           retrySleeper.sleep(sleep);
         }
       }
     }
   }
 
-  private StagingResult tryStagePackage(
-      PackageAttributes attributes, CreateOptions createOptions)
+  private StagingResult tryStagePackage(PackageAttributes attributes, CreateOptions createOptions)
       throws IOException, InterruptedException {
-    File source = attributes.getSource();
+    String sourceDescription = attributes.getSourceDescription();
     String target = attributes.getDestination().getLocation();
 
-    LOG.info("Uploading {} to {}", source, target);
+    LOG.info("Uploading {} to {}", sourceDescription, target);
     try (WritableByteChannel writer =
         FileSystems.create(FileSystems.matchNewResource(target, false), createOptions)) {
-      copyContent(attributes.getSource(), writer);
+      if (attributes.getBytes() != null) {
+        ByteSource.wrap(attributes.getBytes()).copyTo(Channels.newOutputStream(writer));
+      } else {
+        File sourceFile = attributes.getSource();
+        checkState(sourceFile != null, "Attempt to stage to without file or bytes");
```

Comment:
awkward sentence: Attempt to stage to without file or bytes

Reply:
Rephrased.